app/src/main/java/org/mozilla/fenix/searchdialog/SearchDialogFragment.kt

```diff
@@ -324,6 +337,66 @@ class SearchDialogFragment : AppCompatDialogFragment(), UserInteractionHandler {
             })
     }
 
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<String>,
+        grantResults: IntArray
+    ) {
+        when (requestCode) {
+            REQUEST_CODE_CAMERA_PERMISSIONS -> qrFeature.withFeature {
+                context?.let { context: Context ->
+                    it.onPermissionsResult(permissions, grantResults)
+                    if (!context.isPermissionGranted(Manifest.permission.CAMERA)) {
+                        cameraPermissionsDenied = true
+                        dismissAndResetFocus()
+                    }
+                }
+            }
+            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        }
+    }
+
+    private fun dismissAndResetFocus() {
+        view?.qr_scan_button?.isChecked = false
+        toolbarView.view.edit.focus()
+        toolbarView.view.requestFocus()
+    }
+
+    private fun showPermissionsNeededDialog() {
```

Comment:
Maybe for testing purposes it would be good to decouple creating and showing from the same function? And the bulk of the function does seem to be mainly about creating dialog, and it is only showed at the end. If not, some comments would definitely improve the readibility here.